src/core/src/service_interfaces/StatusHandler.py

```diff
@@ -776,10 +796,206 @@ def __try_add_error(error_list, detail):
+
+            self.composite_logger.log("Complete Truncation")
+
+        # Write to status file
+        self.env_layer.file_system.write_with_retry_using_temp_file(self.status_file_path, '[{0}]'.format(json.dumps(truncated_status_file)), mode='w+')
+
+    def __truncate_assessment_helper_func(self, assessment_patches, size_limit):
+        patches_first_half, patches_second_half = (assessment_patches[:5], assessment_patches[5:]) if len(assessment_patches) > 5 else (assessment_patches, [])
+
+        capacity = size_limit - self.__get_byte_size(patches_first_half)
+
+        # Check if the first 5 elements in the list are greater than the size limit
+        assert capacity >= 0, "There are element(s) greater than the size limit"
+
+        new_assessment_list, truncated_packages, _ = self.__apply_truncation(patches_second_half, capacity)
+
+        return patches_first_half + new_assessment_list, truncated_packages
+
+    def __apply_truncation(self, patch_list, size_limit):
+        """ Binary search
+        Instead of checking list[middel_index] >= target, check byte_size(list[:middle_index]),
+        as byte_size[list[:i]] is monotonically increasing, i.e.
+        byte_size[list[:1]] < byte_size[list[:2]] < byte_size[list[:3]] ...
+        return new_patch_list, truncated_packages, and remaining size_limit
+        """
+        left_index = 0
+        right_index = len(patch_list) - 1
+
+        if self.__get_byte_size(patch_list) <= size_limit:
+            return patch_list, [], size_limit - self.__get_byte_size(patch_list)
+        elif self.__get_byte_size(patch_list[0]) > size_limit:
+            return [], patch_list, size_limit
+
+        while left_index < right_index:
+            mid_index = left_index + int((right_index - left_index) / 2)
+            if self.__get_byte_size(patch_list[:mid_index]) >= size_limit:
+                right_index = mid_index
+            else:
+                left_index = mid_index + 1
+
+        new_list = patch_list[:left_index - 1]
+        truncated_packages = patch_list[left_index - 1:]
+        return new_list, truncated_packages, size_limit - self.__get_byte_size(new_list)
+
+    def __read_complete_status_file_raw_data(self, file_path):
+        for i in range(0, Constants.MAX_FILE_OPERATION_RETRY_COUNT):
+            try:
+                with self.env_layer.file_system.open(file_path, 'r') as file_handle:
+                    status_file_data_raw = json.load(file_handle)[0]  # structure is array of 1
+            except Exception as error:
+                if i < Constants.MAX_FILE_OPERATION_RETRY_COUNT - 1:
+                    time.sleep(i + 1)
+                else:
+                    self.composite_logger.log_error(
+                        "Unable to read status file (retries exhausted). Error: {0}.".format(repr(error)))
+                    raise
+        return status_file_data_raw
```

Comment:
Please review if this cannot be done generically - if you have a plan sorted, please discuss before translating into an additional iteration.

Reply:
Look at envlayer's read_with_retry to avoid code duplication. This is a plain vanilla use-case for that.